include/flexible.hxx

```diff
@@ -0,0 +1,464 @@
+    } else {
+      throw BoutException("Not yet implemtented!");
+    }
+    clean(false);
+    return *this;
+  };
+  Flexible<F>& operator/=(BoutReal rhs) {
+    ((F*)fields[mainid])->operator/=(rhs);
+    clean(false);
+    return *this;
+  };
+  Flexible<F>& operator+=(const Field3D & rhs) {
+    if (mainid == getId(rhs.getLocation())){
+      ((F*)fields[mainid])->operator+=(rhs);
+    } else {
+      throw BoutException("Not yet implemtented!");
+    }
+    clean(false);
+    return *this;
+  };
+  Flexible<F>& operator+=(const Field2D & rhs) {
+    if (mainid == getId(rhs.getLocation())){
+      ((F*)fields[mainid])->operator+=(rhs);
+    } else {
+      throw BoutException("Not yet implemtented!");
+    }
+    clean(false);
+    return *this;
+  };
+  Flexible<F>& operator+=(BoutReal rhs) {
+    ((F*)fields[mainid])->operator+=(rhs);
+    clean(false);
+    return *this;
+  };
+  Flexible<F>& operator-=(const Field3D & rhs) {
+    if (mainid == getId(rhs.getLocation())){
+      ((F*)fields[mainid])->operator-=(rhs);
+    } else {
+      throw BoutException("Not yet implemtented!");
+    }
+    clean(false);
+    return *this;
+  };
+  Flexible<F>& operator-=(const Field2D & rhs) {
+    if (mainid == getId(rhs.getLocation())){
+      ((F*)fields[mainid])->operator-=(rhs);
+    } else {
+      throw BoutException("Not yet implemtented!");
+    }
+    clean(false);
+    return *this;
+  };
+  Flexible<F>& operator-=(BoutReal rhs) {
+    ((F*)fields[mainid])->operator-=(rhs);
+    clean(false);
+    return *this;
+  };
+
+private:
+  // Helper function to get index of location.
+  uint getId(CELL_LOC loc_){
+    uint loc = static_cast<uint>(loc_)-1;
+    if ( loc > num_fields || loc < 0){
+      throw BoutException("Unexpeted Fieldlocation! (Info: I got %d - %s)",loc,strLocation(loc_));
+    }
+    return loc;
+  };
+  void init(const F * main){
+    mainloc=main->getLocation();
+    mainid = getId(mainloc);
+    for (uint i=0;i<num_fields;++i){
+      fields[i]=nullptr;
+    }
+    fields[mainid]=main;
+    owner[mainid]=true;
+  };
+  void clean(bool include_main){
+    for (uint i=0;i<num_fields;i++){
+      if (i != mainid || include_main) {
+        if (fields[i] != nullptr){
+	  if (owner[i]){
+	    delete fields[i];
+	  }
+	  fields[i]=nullptr;
+        }
+      }
+    }
+  };
+  const F * geti(CELL_LOC loc_) {
+    if (loc_ == CELL_DEFAULT){
+      return fields[mainid];
+    }
+    uint loc=getId(loc_);
+    if (fields[loc] == nullptr){
+      if (fields[0] == nullptr){
+	fields[0]=new F(interp_to((*fields[mainid]),CELL_CENTRE));
+	owner[mainid]=true;
+      }
+      if (loc != mainid) {
+	fields[loc]=new F(interp_to(*fields[mainid],loc_));
+	owner[mainid]=true;
+      }
+    }
+    ASSERT1(fields[mainid]!=nullptr);
+    return fields[loc];
+  };
+  // Number of field locations we support
+  static const uint num_fields=4;
+  // The pointers to the fields. Some may be null
+  const F * fields[num_fields];
```

Comment:
Why the `const` here? Is `Flexible<F> x` not supposed to modify its fields? I assume this being `const` is why there's casts everywhere.

Reply:
I thought it would be safer to keep const pointers and not modify the fields - but it turns out that does not work so well.
I have changed it to non-const ...